packages/devtools_app/lib/src/screens/memory/panes/leaks/leaks_pane.dart

```diff
@@ -148,122 +143,184 @@ class _LeaksPaneState extends State<LeaksPane>
     final json = jsonEncode(task.toJson());
     final jsonFile = _exportController.generateFileName(
       time: now,
-      prefix: _file_prefix,
+      prefix: _Constants.file_prefix,
       extension: 'raw.json',
     );
     return _exportController.downloadFile(json, fileName: jsonFile);
   }
 
+  void dispose() {
+    status.dispose();
+  }
+}
+
+class _LeaksPaneState extends State<LeaksPane>
+    with
+        AutoDisposeMixin,
+        ProvidedControllerMixin<MemoryController, LeaksPane> {
+  late _LeaksPaneController _controller;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    if (!initController()) return;
+    _controller = _LeaksPaneController(memoryController: controller);
+    cancelStreamSubscriptions();
+    _subscribeForMemoryLeaksMessages();
+  }
+
+  @override
+  void dispose() {
+    _controller.dispose();
+    super.dispose();
+  }
+
   void _subscribeForMemoryLeaksMessages() {
     autoDisposeStreamSubscription(
       serviceManager.service!.onExtensionEventWithHistory.listen((event) async {
-        if (event.extensionKind == _extensionKindToReceiveLeaksSummary) {
-          _receivedLeaksSummary(event);
+        if (event.extensionKind ==
+            _Constants.extensionKindToReceiveLeaksSummary) {
+          _controller.receivedLeaksSummary(event);
         }
       }),
     );
 
     autoDisposeStreamSubscription(
       serviceManager.service!.onExtensionEvent.listen((event) async {
-        if (event.extensionKind == _extensionKindToReceiveLeaksDetails) {
-          await _receivedLeaksDetails(event);
+        if (event.extensionKind ==
+            _Constants.extensionKindToReceiveLeaksDetails) {
+          await _controller.receivedLeaksDetails(event);
         }
       }),
     );
   }
```

Comment:
these can be moved to _LeaksPaneController to make this widget code even more simple. A good sign that code like this should be in the controller is that your listener is calling some action on the controller. Moving this to the controller should also allow us to make this widget stateless

Reply:
done. The widget owns _leaksController, that needs to be disposed, so it is still stateful.